doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md

```diff
@@ -0,0 +1,447 @@
+- Easy to apply per-node configuration during deployment
+- No need to implement a special mechanism for per-node configurations during
+  firmware updates (only in the case of migration)
+- Common interface for modules to expose their runtime configuration parameters
+  and handle them
+- Transparent interface for storing configuration parameters in non-volatile
+  storage devices
+
+## Status
+This document is currently under open discussion. This document is a product of
+[Configuration Task
+Force](https://github.com/RIOT-OS/RIOT/wiki/Configuration-Task-Force-(CTF)), and
+aims to describe the architecture of a runtime configuration system. The content
+of this document is licensed with a Creative Commons CC-BY-SA license.
+
+## Terminology
+This memo uses the [RFC2119](https://www.ietf.org/rfc/rfc2119.txt) terminology
+and the following acronyms and definitions:
+
+### Acronyms
+- RDM: RIOT Developer Memo
+- CTF: RIOT Configuration Task Force
+- RCS: Runtime Configuration System
+- RH: Registry Handler
+- SF: Storage Facility
+
+### Definitions
+- Configuration group: A set of key-value configurations under the same name
+  scheme and scope. E.g `LIGHT_SENSOR_THRESHOLD` and `TRANSMISSION_PERIOD`
+configuration parameters can be contained in an _Application_ configuration
+group, as well as `IEEE802154_CHANNEL` and `IEEE802154_TX_POWER` in a
+_IEEE802.15.4 Radio_ configuration group. Within RIOT, a Configuration Group
+is represented by a Registry Handler.
+
+- Registry Handler: A descriptor that acts as an interface between the RIOT
+Registry and a module that exposes configurations. It provides a common
+interface to get, set, apply and export configurations.
+
+- Storage Facility: A descriptor that acts as an interface between the RIOT
+Registry and a non-volatile storage device. It provides a common interface for
+load/store key-value data from storage devices that might have different
+data representations.
+
+# 1. Introduction
+This document specifies the proposed architecture by the
+Configuration Task Force (CFT) to implement a secure and reliable Runtime
+Configuration System (RCS) focusing on modularity, reuse of existing
+technologies in RIOT (network stack, storage interface) and standard compliance.
+
+# 2. Architecture
+The proposed RCS architecture is formed by a
+[configuration manager](4-configuration-manager) and the
+[RIOT Registry](3-the-riot-registry) and its sub-components:
+one or more [Registry Handlers](4-registry-handlers) and one or more
+[storage facilities](5-storage-facilities). 
+
+A RIOT Application may interact with the Configuration Manager in order to
+modify access control rules or enable different communication interfaces.
+Also, it may interact with the RIOT Registry directly if it's
+required to load/store a persistent configuration.
+
+![](./files/rdm-draft-alamos-lanzieri-runtime-configuration-architecture/architecture.svg "Runtime Configuration Architecture")
```

Comment:
Would be nice if all the arrows went in one direction (eg down)

Reply:
the arrows are there on purpose to show who is the caller. Maybe I should add this?

Reply:
I meant to re-arrange the boxes with the I/O at the top and RHs / SFs at the bottom, so there's some kind of flow to the whole diagram from top to bottom. I could see that the arrows mean caller -> callee, but nevertheless it might be worth putting a label on so it's clear at-a-glance, yes

Reply:
I can add it in the text below the diagram

Reply:
The original comment wasn't addressed, about the layout of the diagram...

Reply:
would that be before or after the PR gets merged? If before, happy to resolve this

Reply:
> would that be before or after the PR gets merged? If before, happy to resolve this

 The diagram has changed.